auth/providers/azure/azure.go

```diff
@@ -80,11 +80,55 @@ type authInfo struct {
 	Issuer      string
 }
 
+// TODO: combine auth info and issuer into one
 var (
 	cachedAuthInfo *authInfo
 	mutex          = &sync.Mutex{}
+
+	cachedOIDCIssuerProvider *oidc.Provider
+	cachedOIDCIssuerProvidersMutex = &sync.RWMutex{}
 )
 
+func getCachedOIDCIssuerProvider() (*oidc.Provider, bool) {
+	cachedOIDCIssuerProvidersMutex.RLock()
+	defer cachedOIDCIssuerProvidersMutex.RUnlock()
+
+	if cachedOIDCIssuerProvider == nil {
+		return nil, false
+	}
+	return cachedOIDCIssuerProvider, true
+}
+
+func getOIDCIssuerProvider(issuerURL string, issuerGetRetryCount int) (*oidc.Provider, error) {
+	// fast path: read from cache
+	if cached, ok := getCachedOIDCIssuerProvider(); ok {
+		return cached, nil
+	}
+
+	// slow path: construct from remote
+	// NOTE: we hold the lock even it's doing HTTP call to avoid sending multiple requests
+	cachedOIDCIssuerProvidersMutex.Lock()
+	defer cachedOIDCIssuerProvidersMutex.Unlock()
+
+	if cached, ok := getCachedOIDCIssuerProvider(); ok {
```

Comment:
This will deadlock.  It tries to take an RLock while a Lock is already held.

Reply:
yes... sorry about the rush, writing a test in local